src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,881 @@
+(** OCaml implementation of the linear two-Variable equalitie domain.
+
+    @see <http://doi.acm.org/10.1145/2049706.2049710> A. Flexeder, M. Petter, and H. Seidl Fast Interprocedural Linear Two-Variable Equalities. *)
+
+(** Abstract states in this domain are represented by structs containing an array and an apron environment.
+    The arrays are modeled as proposed in the paper: Each variable is assigned to an index and each array element represents a linear relationship that must hold at the corresponding program point.
+    The apron environment is hereby used to organize the order of columns and variables.
+*)
+
+open Batteries
+open GoblintCil
+open Pretty
+module M = Messages
+open Apron
+open VectorMatrix
+
+
+
+module Mpqf = SharedFunctions.Mpqf
+
+module Equality = struct
+  (* (Some i, k) represents a sum of a variable with index i and the number k.
+     (None, k) represents the number k. *)
+  type t = (int option * Z.t) [@@deriving eq, ord, hash]
+  let zero = (None, Z.zero)
+  let var_zero i = (Some i, Z.zero)
+  let to_int x = Z.to_int @@ snd x
+  let print : t -> unit = fun (a, b) -> match a with
+    | None -> print_endline @@ "(None , " ^ Z.to_string b ^ ")"
+    | Some x -> print_endline @@ "(Some " ^ string_of_int x ^ ", " ^ Z.to_string b ^ ")"
+end
+
+module EqualitiesArray = struct
+  include Array
+  type t = Equality.t Array.t [@@deriving eq, ord]
+
+  let hash : t -> int = Array.fold_left (fun acc a -> 31 * acc + Equality.hash a) 0
+
+  let empty () = [||]
+
+  let make_empty_array len = Array.mapi (fun i (x, y) -> (Some i, Z.zero)) (make len Equality.zero)
```

Comment:
Maybe `init` is better here, as you replace the initial value everywhere anyway?

Reply:
We would need a `initi` i.e. where the index is available to the initializer, but this is not available in Array or BatArray, but we can use modify to not initialize a whole new array